core/commonMain/src/Options.kt

```diff
@@ -47,155 +56,207 @@ abstract class AbstractSingleOption<T: Any, TResult> internal constructor(owner:
+            ), owner
+        )
     }
     owner.entity = newOption
     return newOption
 }
 
 /**
- * Require option to be always provided in command line.
+ * Requires the option to be always provided in command line.
  */
-fun <T: Any> SingleNullableOption<T>.required(): SingleOption<T> {
+fun <T : Any> SingleNullableOption<T>.required(): SingleOption<T, DefaultRequiredType.Required> {
     val newOption = with((delegate as ParsingValue<T, T>).descriptor as OptionDescriptor) {
-        SingleOption(OptionDescriptor(optionFullFormPrefix, optionShortFromPrefix, type, fullName,
+        SingleOption<T, DefaultRequiredType.Required>(
+            OptionDescriptor(
+                optionFullFormPrefix, optionShortFromPrefix, type, fullName,
                 shortName, description, defaultValue,
-                true, multiple, delimiter, deprecatedWarning), owner)
+                true, multiple, delimiter, deprecatedWarning
+            ), owner
+        )
     }
     owner.entity = newOption
     return newOption
 }
 
 /**
- * Require option to be always provided in command line.
+ * Requires the option to be always provided in command line.
  */
-fun <T: Any, OptionType: MultipleOptionType>
-        MultipleOption<T, OptionType>.required(): MultipleOption<T, OptionType> {
+fun <T : Any, OptionType : MultipleOptionType>
+        MultipleOption<T, OptionType, DefaultRequiredType.None>.required():
+        MultipleOption<T, OptionType, DefaultRequiredType.Required> {
     val newOption = with((delegate as ParsingValue<T, List<T>>).descriptor as OptionDescriptor) {
-        MultipleOption<T, OptionType>(OptionDescriptor(optionFullFormPrefix, optionShortFromPrefix, type, fullName, shortName,
+        MultipleOption<T, OptionType, DefaultRequiredType.Required>(
+            OptionDescriptor(
+                optionFullFormPrefix, optionShortFromPrefix, type, fullName, shortName,
                 description, defaultValue?.toList() ?: listOf(),
-                true, multiple, delimiter, deprecatedWarning), owner)
+                true, multiple, delimiter, deprecatedWarning
+            ), owner
+        )
     }
     owner.entity = newOption
     return newOption
 }
 
 /**
- * Allow provide several options using [delimiter].
+ * Allows the option to have several values joined with [delimiter] specified in command line string.
+ * Number of values is unlimited.
  *
- * @param delimiterValue delimiter used to separate string value to option values.
+ * The value of the argument is an empty list in case if no value was specified in command line string.
+ *
+ * @param delimiterValue delimiter used to separate string value to option values list.
  */
-fun <T : Any, TResult> AbstractSingleOption<T, TResult>.delimiter(delimiterValue: String): MultipleOption<T, DelimitedOption> {
+fun <T : Any, TResult, DefaultRequired: DefaultRequiredType> AbstractSingleOption<T, TResult, DefaultRequired>.delimiter(
```

Comment:
`TResult` seems to be ignored. Is it ok here?

Reply:
It's needed for type inference.